### DISCUSSION TITLE
Curso " Fundamentos da Orientação a Objetos" (Evento e delegação)

### DISCUSSION THREAD

**Discussion by rogerfsp** [balta-io/forum#401] [Tira Dúvidas] [answered]
C#: Considerando o código a seguir, por que a linha **"vRoom.RoomSoldOutEvent += OnRoomSoldOut;"**, está errada?:

_ERRO =  error CS0123: Nenhuma sobrecarga de "OnRoomSoldOut" corresponde ao representante "EventHandler"_

namespace Events01
{
    class Program
    {
        static void Main(string[] args)
        {
            var vRoom = new Room(3);
            vRoom.RoomSoldOutEvent += OnRoomSoldOut;  // apresentando erro
            vRoom.ReserveSeat();
            vRoom.ReserveSeat();
            vRoom.ReserveSeat();
            vRoom.ReserveSeat();
            vRoom.ReserveSeat();
            vRoom.ReserveSeat();
        }
        static void OnRoomSoldOut()
        {
            Console.WriteLine("Sala lotada.");
        }
    }

    public class Room
    {
        private int seatsInUse = 0;
        public int Seats { get; set; }

        //-- CONSTRUTOR -----------------------------------------------------------------------------------------------
        public Room(int pSeats)
        {
            Seats = pSeats;
            seatsInUse = 0;
        }

        public void ReserveSeat()
        {
            seatsInUse++;
            if (seatsInUse >= Seats)
            {
                OnRoomSoldOut(EventArgs.Empty);
            }
            else
            {
                Console.WriteLine("Assento reservado");
            }
        }

        public event EventHandler RoomSoldOutEvent;  // DEFINIÇÃO DO EVENTO

        protected virtual void OnRoomSoldOut(EventArgs e) // EXECUÇÃO DO EVENTO
        {
            EventHandler handler = RoomSoldOutEvent;
            handler?.Invoke(this, e);
        }
    }
}

**Comment by Romulo-Queiroz** [comment]
Boa noite @rogerfsp 
Poderia nos enviar uma imagem do código?
Se possível inclua a imagem da exceção por favor, assim nós do fórum iremos conseguir te auxiliar melhor.

**Comment by rogerfsp** [comment]
Olá Rômulo,

Antes de mais nada, agradeço a atenção. Sou programador há 300 anos, mas há
pouco tempo comecei a fazer cursos de C#. Acabei não gostando de nenhum e
agora estou tentando o Balta.

Segue a imagem (ANEXO), do que acredito que seja o que me solicitou. Por
favor analise o problema para mim. Acabei parando nesse tópico, o que está
me impedindo de avançar, sem entendê-lo.

Muito obrigado.

*Roger Fernandes*
*DEV MEMORA Processos Inovadores*
***@***.*** ***@***.***>*
*24 99958 9105*

Em sex., 22 de set. de 2023 às 19:44, Rômulo Queiroz <
***@***.***> escreveu:

**Comment by rogerfsp** [comment]
@Romulo-Queiroz , respondi pelo e-mail (com anexo). Segue aqui também, link da imagem:  https://drive.google.com/file/d/1AbvrfOC8g5CZ6_-GPYm5_dm8phMdXIvz/view?usp=sharing

**Reply by andrebaltieri** [reply, marked as answer]
Acredito que seja o `static` no método `OnRoomSoldOut`. Tenta remover o `static` e veja se funciona.

**Comment by rogerfsp** [comment]
Olá @andrebaltieri , infelizmente retirando o static do método OnRoomSoldOut, ainda não resolveu, e o erro persiste¹. Tanto no VS Code versão 1.82.2, quanto no Visual Studio 2022 (versão 17.7.3), o erro é o mesmo. Por favor, analise para mim, acabei parando neste ponto. Muito obrigado.

¹ _No overload for 'OnRoomSoldOut' matches delegate 'EventHandler' (CS0123)_

( Observação: no seu código, no vídeo, não ocorre o erro )

**Comment by rogerfsp** [comment]
Inserindo:  object sender, EventArgs e,  como parâmetros no método, corrige e o programa executa :

static void OnRoomSoldOut (object sender, EventArgs e)  // inseridos parâmetros
        {
            Console.WriteLine("Sala lotada.");
        }

Analise isto, por favor, obrigado!

**Reply by andrebaltieri** [reply]
Acredito que este é o código apresentado no curso, não?

**Reply by rogerfsp** [reply]
Você tem razão. Perdão, mas, ficou uma marcação warning (sublinhado amarelo) no OnRoomSoldOut, na linha  vRoom.RoomSoldOutEvent += OnRoomSoldOut;  .  Com este aviso:
Nullability of reference types in type of parameter 'sender' of 'void Program.OnRoomSoldOut(object sender, EventArgs e)' doesn't match the target delegate 'EventHandler' (possibly because of nullability attributes)

Em "  public Room(int pSeats)",   Room também está sublinhado amarelo:

Non-nullable event 'RoomSoldOutEvent' must contain a non-null value when exiting constructor. Consider declaring the event as nullable.

**Reply by andrebaltieri** [reply]
@rogerfsp aqui tem um material sobre:
https://go.balta.io/ebook-nullable-types-dotnet?utm_source=GitHub
